agent/agentcontainers/api.go

```diff
@@ -544,6 +549,65 @@ func (api *API) Routes() http.Handler {
 	return r
 }
 
+func (api *API) watchContainers(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	conn, err := websocket.Accept(rw, r, nil)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to upgrade connection to websocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	ctx = api.ctx
+
+	go httpapi.Heartbeat(ctx, conn)
+	defer conn.Close(websocket.StatusNormalClosure, "connection closed")
+
+	encoder := wsjson.NewEncoder[codersdk.WorkspaceAgentListContainersResponse](conn, websocket.MessageText)
+	defer encoder.Close(websocket.StatusNormalClosure)
+
+	updateCh := make(chan struct{}, 1)
+	defer func() {
+		api.mu.Lock()
+		close(updateCh)
```

Comment:
It'd make sense to move this to the defer below to keep the related code close to each other.